### PR TITLE
`driver`: CIP-20 support

### PR DESCRIPTION
**Pull request by cowfee**
Adds CIP-20 support into https://github.com/cowprotocol/services/pull/1425 by @sunce86. Updates the tests to ensure that generated scores are always positive, negative-scoring settlements get filtered out, and the new API changes are respected.

Those tests which aren't using the nice new test setup framework got commented out temporarily, but I'll make it a high priority to fix these as soon as I'm done with my work on `solvers`. The related issue is [here](https://github.com/cowprotocol/services/issues/1527).

### Test Plan

Automated tests and faith in myself.

### CODE REVIEW

**Comment by cowfee** [comment]
There's a flaky test, moving to draft.

**Comment by cowfee** [comment]
There's a big inconsistency in the test setup, this is not ready for review yet.

**Review by vkgnosis — APPROVED** [review]

File: crates/driver/src/boundary/settlement.rs [hunk 1]
```diff
@@ -213,6 +214,7 @@ impl Settlement {
             gas_price,
             &gas.into(),
         );
+        ensure!(!inputs.objective_value().is_negative(), "negative score",);
```
Comment (nlordell) [inline crates/driver/src/boundary/settlement.rs:217]:
```suggestion
        ensure!(!inputs.objective_value().is_negative(), "negative score");
```

`rustfmt` doesn't remove trailing commas in macro invocations 🤷.

File: crates/driver/src/domain/competition/mod.rs [hunk 2]
```diff
@@ -46,20 +45,20 @@ pub struct Competition {
     pub settlement: Mutex<Option<Settlement>>,
 }
 
+/// Solution information revealed to the protocol by the solver. Note that the
+/// calldata is never revealed, as it would allow the solution to be stolen by
+/// other solvers.
```
Comment (nlordell) [inline crates/driver/src/domain/competition/mod.rs:50]: **Not for this PR**, but I believe CIP-20 stipulates that solvers need to reveal calldata (well, specifically, a "solution" which can be turned into calldata). We _may_ want to change the driver co-location API to accommodate this. @harisang - what are your thoughts?

File: crates/driver/src/infra/solver/dto/solution.rs [hunk 3]
```diff
@@ -220,7 +219,6 @@ pub struct Solution {
     prices: HashMap<eth::H160, eth::U256>,
     trades: Vec<Trade>,
     interactions: Vec<Interaction>,
-    score: Option<eth::U256>,
```
Comment (nlordell) [inline crates/driver/src/infra/solver/dto/solution.rs:223]: In CIP-20, the solver engines provide a "risk" value to the solution they propose (called "score discount" in the CIP AFAIR). This flat value is subtracted from the driver-computed score and represents how much risk the solver thinks the transaction has.

I'm fine with including this in a separate PR, but it is needed for full CIP-20 support.

**Review by nlordell — APPROVED** [review]
Changes look good to me. Two comments that I think require follow-ups.

File: crates/driver/src/infra/solver/dto/solution.rs (hunk 3 above)
Comment (cowfee) [reply crates/driver/src/infra/solver/dto/solution.rs:223]: https://github.com/cowprotocol/services/pull/1526/commits/c94cfe099feecf2f1eaf3559a5f822823ada790f

File: crates/driver/src/boundary/settlement.rs (hunk 1 above)
Comment (cowfee) [reply crates/driver/src/boundary/settlement.rs]: https://github.com/cowprotocol/services/pull/1526/commits/d70391d60e725367c3b95905ceaa0344f51e0040

File: crates/driver/src/infra/solver/mod.rs [hunk 4]
```diff
@@ -60,10 +60,22 @@ pub struct Config {
...
+pub struct Rewards {
+    /// Address to which performance rewards will be sent. Performance rewards
+    /// are given out whenever the solver submits a winning solution.
+    pub performance: eth::Address,
+    /// Address to which participation rewards will be sent. Participation
+    /// rewards are given out whenever the solver submits a solution, regardless
+    /// of whether or not it wins.
+    pub participation: eth::Address,
+}
+
```
Comment (nlordell) [inline crates/driver/src/infra/solver/mod.rs:78]: This isn't used anywhere right?

**Pull request merged**